coap-gateway: handle multiple uses of auth code when retrying signUp

Danielius1922 opened the pull request.

Danielius1922 committed:
pkg: create Future type

Type Future allows multiple gorouties to block and wait
for a value to be set by a producer.

Danielius1922 committed:
test/oauth-server: ClientTestRestrictedAuth client

When oauth-server is used with ClientTestRestrictedAuth client,
the server keeps track of used auth codes and refresh tokens.
When the same auth code or refresh token is used repeatedly within
a minute of the first use, the repeated uses return an error.

jkralik commented on pkg/sync/task/future/future.go:
```diff
@@ -0,0 +1,67 @@
+	default:
+		f.value = value
+		close(f.ready)
+	}
```
this is not threadsafe -> if 2 goroutines are called at the same time both of them can be at line 44

Danielius1922 replied:
set is returned from `New` only, so it's accessible in the goroutine it is constructed in

Danielius1922 replied:
Hmm, `oauth2.Token` is a value type, so I don't actually need this type, I can just use channels to wait for data and channels are always thread-safe, so that would be better.

jkralik commented on pkg/sync/task/future/future.go:
```diff
@@ -0,0 +1,67 @@
+	default:
+		f.err = err
+		close(f.ready)
+	}
```
same as previous comment

jkralik commented on coap-gateway/service/exchange.go:
```diff
@@ -0,0 +1,100 @@
+		}
+		e.codeToData[authorizationCode] = data
+	}
+	e.lock.Unlock()
```
I prefer to use `defer e.lock.Unlock()` so I suggest moving the whole code under lock to a separate function

jkralik commented on coap-gateway/service/exchange.go:
```diff
@@ -0,0 +1,99 @@
+			setToken(nil, err)
+			return oauth2.Token{}, err
+		}
+		if err := e.setTokenForCode(authorizationCode, token.AccessToken.String()); err != nil {
```
why you store `token.AccessToken.String()` ?

jkralik commented on coap-gateway/service/signIn.go:
```diff
@@ -316,6 +316,9 @@ func signInPostHandler(req *mux.Message, client *Client, signIn CoapSignInReq) {
 		client.server.expirationClientCache.Set(deviceID, client, time.Second*time.Duration(expiresIn))
 	}
+
+	client.exchangeHandler.RemoveDataForToken(signIn.AccessToken)
```
drop all tokens which are stored at the client. the device never tries signUp through the same connection again when it starts to sign in.

jkralik commented:
We need to implement the same feature for the refresh token because refreshing token can occurs only once for the same refresh token. (depends on OAuth server)

jkralik commented on coap-gateway/service/refreshToken.go:
```diff
@@ -71,24 +71,24 @@ func validUntilToExpiresIn(validUntil time.Time) int64 {
+		errSubmit := client.server.taskQueue.Submit(func() {
 			defer wg.Done()
-			tokenTmp, errTmp := provider.Refresh(ctx, refreshToken)
+			tokenTmp, errTmp := client.tokenCache.Refresh(ctx, provider, refreshToken)
```
This can be called from multiple threads so `client.tokenCache.Refresh` must be thread-safe.

Danielius1922 replied:
The `refreshTokenPostHandler` is not serialized like the `signUpPostHandler`?

jkralik commented on coap-gateway/service/tokenCache.go:
```diff
@@ -0,0 +1,83 @@
+	}
+	if refreshToken == e.refresh.code {
+		return e.refresh.token, nil
+	}
```
need to be thread safe.

jkralik commented on coap-gateway/service/tokenCache.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	e.refresh.code = refreshToken
+	e.refresh.token = *token
```
same here

ondrejtomcik commented:
@jkralik 100 lines?

Danielius1922 committed:
coap-gateway: create exchangeCache

OAuth server returns error if the same auth code is used
multiple times to request an access token. Exchange cache
stores the last used (code, token) pair. Repeated use of
the same code value does not trigger sending of requests to
server and instead returns the stored token.

Danielius1922 committed:
coap-gateway: avoid repeated Exchange calls

Use exchangeCache in signUp handler to avoid sending
multiple Exchange requests to oauth server with the same
authorization code value.

Danielius1922 committed:
coap-gateway: create refreshCache

OAuth server returns error if the same refresh token is used
multiple times to request an access token. Refresh cache stores
the last used (refreshToken, token) pair. Repeated use of the
same code value does not trigger sending of requests to server
and instead returns the stored token.

Danielius1922 committed:
coap-gateway: avoid repeated Refresh calls

Use refreshCache in refreshToken handler to avoid sending
multiple Refresh requests to oauth server with the same
refresh token.

jkralik commented on coap-gateway/service/signIn.go:
```diff
@@ -316,6 +316,9 @@ func signInPostHandler(req *mux.Message, client *Client, signIn CoapSignInReq) {
 	}
+
+	client.exchangeCache.Clear()
+
```
we want to clear refresh token after success sign in - `client.refreshCache.Clear()`

jkralik commented:
> @jkralik 100 lines?

Ok, I can leave with 600+ lines 😸

Danielius1922 commented:
> > @jkralik 100 lines?
> 
> Ok, I can leave with 600+ lines smile_cat

If we ignore the lines in tests then it's about 100 lines. :D
